File: Template Pattern/Rollenspiel_cpp/ASpielfigur.cpp
#include <iostream>

#include "ASpielfigur.h"

ASpielfigur::ASpielfigur(AWaffenverhalten* wv)
{
	this->set_waffe(wv);
}

ASpielfigur::~ASpielfigur()
{
	delete this->wv_;
}

auto ASpielfigur::anzeigen() -> void
{
	std::cout << "Ich bin eine Spielfigur vom Typ: " << this->get_class_name() << "\n";
}

auto ASpielfigur::kaempfen(unsigned const int& anzahl) -> void
{
	this->get_waffe()->benutzeWaffe(anzahl);
}

auto ASpielfigur::set_waffe(AWaffenverhalten* wv) -> void
{
	delete this->wv_;
	this->wv_ = wv;
}

auto ASpielfigur::get_waffe() -> AWaffenverhalten*
{
	return this->wv_;
}

auto ASpielfigur::get_class_name() -> std::string
{
	return std::string(typeid(*this).name()).substr(6, std::string::npos);
}

File: Template Pattern/Rollenspiel_java/src/main/Spielfigur/ASpielfigur.java
package main.Spielfigur;

import main.Waffenverhalten.AWaffenverhalten;

public abstract class ASpielfigur {
    private AWaffenverhalten waffe;

    public ASpielfigur(final AWaffenverhalten wv) {
        this.setWaffe(wv);
    }

    public final void anzeigen() {
        System.out.print("Ich bin eine Spielfigur vom Typ: ");
        System.out.println(this.getClassName());
    }

    public final void kaempfen(final int anzahl) {
        this.getWaffe().benutzeWaffe(anzahl);
    }

    private AWaffenverhalten getWaffe() {
        return this.waffe;
    }

    public final void setWaffe(final AWaffenverhalten wv) {
        if (wv != null) {
            this.waffe = wv;
        } else {
            throw new IllegalArgumentException();
        }
    }

    private String getClassName() {
        var name = this.getClass().getName().split("\\.");
        return name[name.length - 1];
    }
}
